packages/blade/src/components/Dropdown/useDropdown.ts

```diff
@@ -0,0 +1,233 @@
+  setSelectedIndices: (value: number[]) => void;
+  options: OptionsType;
+  setOptions: (value: OptionsType) => void;
+  activeIndex: number;
+  setActiveIndex: (value: number) => void;
+  shouldIgnoreBlur: boolean;
+  setShouldIgnoreBlur: (value: boolean) => void;
+  dropdownBaseId: string;
+  selectInputRef: {
+    current: HTMLButtonElement | null;
+  };
+  actionListRef: {
+    current: HTMLDivElement | null;
+  };
+  selectionType?: DropdownProps['selectionType'];
+};
+
+const DropdownContext = React.createContext<DropdownContextType>({
+  isOpen: false,
+  setIsOpen: noop,
+  selectedIndices: [],
+  setSelectedIndices: noop,
+  options: [],
+  setOptions: noop,
+  activeIndex: -1,
+  setActiveIndex: noop,
+  shouldIgnoreBlur: false,
+  setShouldIgnoreBlur: noop,
+  dropdownBaseId: '',
+  actionListRef: {
+    current: null,
+  },
+  selectInputRef: {
+    current: null,
+  },
+});
+
+let searchTimeout: number;
+// eslint-disable-next-line one-var
+let searchString = '';
+
+type UseDropdownReturnValue = DropdownContextType & {
+  onSelectClick: React.MouseEventHandler<HTMLInputElement>;
+  onSelectKeydown: FormInputHandleOnKeyDownEvent | undefined;
+  onSelectBlur: FormInputHandleOnEvent | undefined;
+  onOptionClick: (
+    e: React.MouseEvent<HTMLButtonElement> | React.KeyboardEvent<HTMLInputElement>,
+    index: number,
+  ) => void;
+  value: string;
+  /**
+   * This is the value that is displayed inside select after selection
+   */
+  displayValue: string;
+};
+const useDropdown = (): UseDropdownReturnValue => {
+  const {
+    isOpen,
+    setIsOpen,
+    selectedIndices,
+    setSelectedIndices,
+    activeIndex,
+    setActiveIndex,
+    shouldIgnoreBlur,
+    setShouldIgnoreBlur,
+    options,
+    selectionType,
+    ...rest
+  } = React.useContext(DropdownContext);
+
+  const selectOption = (
+    index: number,
+    properties: {
+      closeOnSelection?: boolean;
+    } = {
+      closeOnSelection: true,
+    },
+  ): void => {
+    if (index < 0 || index > options.length - 1) {
+      return;
+    }
+
+    if (selectionType === 'multiple') {
+      if (selectedIndices.includes(index)) {
+        // remove existing item
+        const existingItemIndex = selectedIndices.indexOf(index);
+        setSelectedIndices([
+          ...selectedIndices.slice(0, existingItemIndex),
+          ...selectedIndices.slice(existingItemIndex + 1),
+        ]);
+      } else {
+        setSelectedIndices([...selectedIndices, index]);
+      }
+    } else {
+      setSelectedIndices([index]);
+    }
+
+    if (activeIndex !== index) {
+      setActiveIndex(index);
+    }
+
+    if (properties?.closeOnSelection && selectionType !== 'multiple') {
+      setIsOpen(false);
+    }
+  };
+
+  const onSelectClick: React.MouseEventHandler<HTMLInputElement> = (_e) => {
+    setIsOpen(!isOpen);
+  };
+
+  const onSelectBlur = (): void => {
+    if (shouldIgnoreBlur) {
+      setShouldIgnoreBlur(false);
+      return;
+    }
```

Comment:
The logic in all these functions here is replicated from w3.org's example (I converted it from vanilla javascript to more react-friendly code).